android/greengrass-nucleus-android/src/main/java/com/aws/greengrass/util/platforms/android/AndroidPlatform.java

```diff
@@ -0,0 +1,727 @@
+    private final AndroidRunWithGenerator runWithGenerator;
+
+    private AndroidServiceLevelAPI androidServiceLevelAPI;
+    private AndroidPackageManager androidPackageManager;
+    private AndroidComponentManager androidComponentManager;
+
+    /**
+     * Construct a new instance.
+     */
+    public AndroidPlatform() {
+        super();
+        runWithGenerator = new AndroidRunWithGenerator(this);
+    }
+
+
+    /**
+     * Set reference to Android Service Level interface to future references.
+     */
+    public void setAndroidAPIs(final AndroidServiceLevelAPI androidServiceLevelAPI,
+                               final AndroidPackageManager androidPackageManager,
+                               final AndroidComponentManager androidComponentManager) {
+        this.androidServiceLevelAPI = androidServiceLevelAPI;
+        this.androidPackageManager = androidPackageManager;
+        this.androidComponentManager = androidComponentManager;
+    }
+
+    /**
+     * Run the `id` program which returns user and group information about a particular user.
+     *
+     * @param id       the identifier (numeric or name). If empty, then the current user is looked up.
+     * @param option   whether to load group or user information.
+     * @param loadName whether a name should or id should be returned.
+     * @return the output of id (either an integer string or name of the user/group) or empty if an error occurs.
+     */
+    private static Optional<String> id(String id, AndroidPlatform.IdOption option, boolean loadName) {
+        boolean loadSelf = Utils.isEmpty(id);
+        String[] cmd = new String[2 + (loadName ? 1 : 0) + (loadSelf ? 0 : 1)];
+        int i = 0;
+        cmd[i] = "id";
+        switch (option) {
+            case Group:
+                cmd[++i] = "-g";
+                break;
+            case User:
+                cmd[++i] = "-u";
+                break;
+            default:
+                logger.atDebug().setEventType("id-lookup")
+                        .addKeyValue("option", option)
+                        .log("invalid option provided for id");
+                return Optional.empty();
+        }
+        if (loadName) {
+            cmd[++i] = "-n";
+        }
+        if (!loadSelf) {
+            cmd[++i] = id;
+        }
+
+        logger.atTrace().setEventType("id-lookup").kv("command", String.join(" ", cmd)).log();
+        StringBuilder out = new StringBuilder();
+        StringBuilder err = new StringBuilder();
+
+        Throwable cause = null;
+        try (Exec exec = getInstance().createNewProcessRunner()) {
+            Optional<Integer> exit = exec.withExec(cmd).withShell().withOut(out::append).withErr(err::append).exec();
+            if (exit.isPresent() && exit.get() == 0) {
+                return Optional.of(out.toString().trim());
+            }
+        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)